changelog/headers.dd

```diff
@@ -1,12 +1,12 @@
 Added experimental `C++` header generation from `extern(C|C++)` declarations
 
-DMD can now write C++ header files which contain bindings for declarations
+DMD can now write C or C++ header files which contain bindings for declarations
 in existing D files that were marked as `extern(C)` or `extern(C++)`.
 
 This feature is available via the following command line switches:
-- `-HC`: Write the generated headers to standard output
-- `-HCf=<file>`: Write the generated headers to file
-- `-HCd=<directory>`: Write the generated headers to file (not implemented yet)
+- `-Hc`, `-HC`: Write the generated headers to standard output (in C or C++ mode, respectively)
+- `-Hcf=<file>`, `-HCf=<file>`: Write the generated headers to file
+- `-Hcd=<directory>`, `-HCd=<directory>`: Write the generated headers to file in the given directory (not implemented yet)
```

Comment:
I don't actually think this is necessary; I think we can emit compatible headers without selecting variants.
I planned to wrap the C++ symbols in `#if defined(__cplusplus)`, then they will be invisible to the C compiler, and make sure the C symbols are C compatible...

Reply:
Probably a better idea.  How does that affect `nullptr`/`_d_null`?

Reply:
`nullptr` can be defined to `NULL` if `__cplusplus < magic_number`.

Reply:
So my plan was to split `decl` into `cdecl` and `cppdecl`, and emit the symbols to one or the other appropriately, and then dump `cppdecl` below wrapped in the ifdef.

Reply:
> `nullptr` can be defined to `NULL` if `__cplusplus < magic_number`.

I don't like that, mainly because it leaks c++ language minutiae into c code.

Reply:
I'm not sure what you mean? `#define` is very much a C thing.
Or if you mean just the name `nullptr`; I think it's safe to say that something close to 100% of use cases will be C++11... `extern(C++)` has some pretty strong affinities for C++11 these days. The only known case of wanting to use C++98 is bootstrapping GDC.

Reply:
I mean, the alternative would be to use `NULL`, but I think that's a whole lot crappier personally.

Reply:
I mean, if a c file includes the (dual) header, it will get a copy of `nullptr`.

Reply:
If that's a real concern, you can undef it before the end of the file.
It's also possible to leave that define to the including code. They may need to specify definitions of the enum macros too, so it's not unreasonable to expect them to `#define nullptr NULL` if we don't want to emit that ourselves.